python/ql/src/semmle/python/dataflow/new/internal/TaintTrackingPrivate.qll

```diff
@@ -148,54 +149,47 @@ predicate stringManipulation(DataFlow::CfgNode nodeFrom, DataFlow::CfgNode nodeT
 predicate containerStep(DataFlow::CfgNode nodeFrom, DataFlow::Node nodeTo) {
   // construction by literal
   // TODO: Not limiting the content argument here feels like a BIG hack, but we currently get nothing for free :|
-  storeStep(nodeFrom, _, nodeTo)
+  DataFlowPrivate::storeStep(nodeFrom, _, nodeTo)
   or
   // constructor call
-  exists(CallNode call | call = nodeTo.asCfgNode() |
-    call.getFunction().(NameNode).getId() in [
-        "list", "set", "frozenset", "dict", "defaultdict", "tuple"
-      ] and
-    call.getArg(0) = nodeFrom.getNode()
+  exists(DataFlow::CallCfgNode call | call = nodeTo |
+    call = API::builtin(["list", "set", "frozenset", "dict", "tuple"]).getACall() and
+    call.getArg(0) = nodeFrom
+    // TODO: Properly handle defaultdict/namedtuple
   )
   or
   // functions operating on collections
-  exists(CallNode call | call = nodeTo.asCfgNode() |
-    call.getFunction().(NameNode).getId() in ["sorted", "reversed", "iter", "next"] and
-    call.getArg(0) = nodeFrom.getNode()
+  exists(DataFlow::CallCfgNode call | call = nodeTo |
+    call = API::builtin(["sorted", "reversed", "iter", "next"]).getACall() and
+    call.getArg(0) = nodeFrom
   )
   or
   // methods
-  exists(CallNode call, string name | call = nodeTo.asCfgNode() |
+  exists(DataFlow::CallCfgNode call, string name | call = nodeTo |
     name in [
         // general
         "copy", "pop",
         // dict
         "values", "items", "get", "popitem"
       ] and
-    call.getFunction().(AttrNode).getObject(name) = nodeFrom.asCfgNode()
+    call.getFunction().(DataFlow::AttrRead).getObject(name) = nodeFrom
```

Comment:
This feels like something we would want to capture directly on `CallCfgNode`. That is, a way of saying "this is a method call with such and such object". 

JavaScript does this using the [`calls`](https://github.com/github/codeql/blob/main/javascript/ql/src/semmle/javascript/dataflow/Nodes.qll#L291) method on `MethodCallNode`. (We may also want to align on these subclasses.)

Reply:
I think in general doing it in this very syntactical way, like this code does, is fundamentally wrong, since it does not capture `meth = my_set.pop; meth()`. 

So while I like the idea of doing this in an easy way, I think we should make it easy to do it the _right_ way instead :wink:

Reply:
> So while I like the idea of doing this in an easy way, I think we should make it easy to do it the _right_ way instead

I may be misreading this, but it seems to be implying that I'm suggesting something that's "an easy way" and not "the _right_ way", but that's really not the case. What I'm suggesting is something that captures "this is a call, and the local source of the function part is an `AttrRead` with such-and-such object and attribute name". This would capture your two-step method call just fine.

Reply:
Very likely that I have misunderstood then. Happy to see a PR for improving this in the future :wink:

Reply:
So what I had in mind is here: https://github.com/github/codeql/pull/6064
... but I see now that this is perhaps not the most appropriate for this particular use case, since it operates at the level of local source nodes. It's likely to still be of some use, though (as we have many instances of "manual method matching" in our various models).